subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectConfigurationReportingTaskExecutionGraph.kt

```diff
@@ -0,0 +1,245 @@
+        }
+        if (otherProjects.isNotEmpty()) {
+            reportCrossProjectTaskAccess(otherProjects)
+        }
+    }
+
+    private
+    val Project.isReferrerProject: Boolean
+        get() = this is ProjectInternal && identityPath == referrerProject.identityPath
+
+    private
+    fun reportCrossProjectTaskAccess(coupledProjects: Iterable<ProjectInternal>, requestPath: String? = null) {
+        reportCoupledProjects(coupledProjects)
+        reportProjectIsolationProblem(requestPath)
+    }
+
+    private
+    fun reportCoupledProjects(coupledProjects: Iterable<ProjectInternal>) {
+        coupledProjects.forEach { other ->
+            coupledProjectsListener.onProjectReference(referrerProject.owner, other.owner)
+        }
+    }
+
+    private
+    fun reportProjectIsolationProblem(requestPath: String?) {
+        val location = userCodeContext.location(null)
+        val message = StructuredMessage.build {
+            text("Project ")
+            reference(referrerProject.identityPath.toString())
+            text(" cannot access the tasks in the task graph that were created by other projects")
+        }
+        val exception = InvalidUserCodeException(
+            // As the exception message is not used for grouping, we can safely add the exact task name to it:
+            message.toString().capitalized() +
+                if (requestPath != null) "; tried to access '$requestPath'" else '"'
+        )
+        problems.onProblem(PropertyProblem(location, message, exception, null))
+    }
+
+    private
+    fun TaskExecutionGraphListener.wrap() = CrossProjectAccessTrackingTaskExecutionGraphListener(this, referrerProject, crossProjectModelAccess)
+
+    private
+    class CrossProjectAccessTrackingTaskExecutionGraphListener(
+        private val delegate: TaskExecutionGraphListener,
+        private val referrerProject: ProjectInternal,
+        private val crossProjectModelAccess: CrossProjectModelAccess
+    ) : TaskExecutionGraphListener {
+
+        override fun graphPopulated(graph: TaskExecutionGraph) {
+            val wrappedGraph = crossProjectModelAccess.taskGraphForProject(referrerProject, graph as TaskExecutionGraphInternal)
+            delegate.graphPopulated(wrappedGraph)
+        }
+
+        override fun equals(other: Any?): Boolean =
+            (other as? CrossProjectAccessTrackingTaskExecutionGraphListener)?.javaClass == javaClass &&
+                other.delegate == delegate &&
+                other.referrerProject == referrerProject
+
+        override fun hashCode(): Int = Objects.hash(delegate, referrerProject)
+        override fun toString(): String = "CrossProjectAccessTrackingTaskExecutionGraphListener(delegate = $delegate, referrerProject = $referrerProject)"
+    }
+
+    private
+    fun Action<in TaskExecutionGraph>.wrap(): Action<TaskExecutionGraph> = Action {
+        val wrappedGraph = if (this@Action == this@CrossProjectConfigurationReportingTaskExecutionGraph)
```

Comment:
❓ What's going on here?

Reply:
The initial idea here was to avoid creating unnecessary new instances of `CrossProjectConfigurationReportingTaskExecutionGraph` if we know that what is passed is already a wrapped task graph.

However, I can now see that this condition is never satisfied because the underlying implementation in `DefaultTaskExecutionGraph` passes itself as the receiver.

I've simplified this in https://github.com/gradle/gradle/pull/22445/commits/ad713bb5852b86466171faec4f993ea8e9088a2b.